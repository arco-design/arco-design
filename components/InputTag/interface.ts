import { CSSProperties, ReactNode } from 'react';

export type ObjectValueType = {
  value?: any;
  label?: ReactNode;
  closable?: boolean;
};

export type ValueChangeReason = 'add' | 'remove' | 'clear' | 'sort';

/**
 * @title InputTag
 */
export interface InputTagProps<T = any> {
  className?: string | string[];
  style?: CSSProperties;
  /**
   * @zh 不同尺寸
   * @en Different sizes
   */
  size?: 'mini' | 'small' | 'default' | 'large';
  /**
   * @zh 预设文案
   * @en Placeholder of input element
   */
  placeholder?: string;

  /**
   * @zh 是否是错误状态.(废弃，下个大版本移除，使用 status='error' 替代)
   * @en Whether the input is error.(Deprecated, removed in the next major version, use status='error' instead)
   * @deprecated status="error"
   */
  error?: boolean;
  /**
   * @zh 状态
   * @en Status
   * @version 2.45.0
   */
  status?: 'error' | 'warning';
  /**
   * @zh 是否禁用
   * @en Whether the input is disabled
   */
  disabled?: boolean;
  /**
   * @zh 自动聚焦
   * @en Whether the input is focused by default
   */
  autoFocus?: boolean;
  /**
   * @zh 是否只读
   * @en Whether the input is read only
   */
  readOnly?: boolean;
  /**
   * @zh 是否允许清除
   * @en Whether to allow clear input value
   */
  allowClear?: boolean;
  /**
   * @zh 是否为内部标签变化添加动画。
   * @en Whether to add animation for internal label changes
   * @version 2.15.0
   * @defaultValue true
   */
  animation?: boolean;
  /**
   * @zh 是否在失焦时自动存储正在输入的文本
   * @en Whether to automatically store the text entering when blur InputTag
   * @version 2.25.0
   */
  saveOnBlur?: boolean;
  /**
   * @zh 默认值
   * @en To set default value
   */
  defaultValue?: T[];
  /**
   * @zh 控件值
   * @en To set value
   */
  value?: T[];
  /**
   * @zh 控件的输入框内的值
   * @en To set input value
   */
  inputValue?: string;
  /**
   * @zh 设置传入和回调出的值均为 `{ label: '', value: ''}` 格式。
   * @en If true, the incoming and callback values will be `{label: '', value: '')` format
   */
  labelInValue?: boolean;
  /**
   * @zh 是否可以通过拖拽为 Tag 排序
   * @en Weather it is possible to sort tags by drag
   * @version 2.27.0
   */
  dragToSort?: boolean;
  /**
   * @zh 添加前缀文字或者图标
   * @en The prefix icon or text for the input-tag
   * @version 2.47.0
   */
  prefix?: ReactNode;
  /**
   * @zh 后缀
   * @en The suffix for the InputTag
   */
  suffix?: ReactNode;
  /**
   * @zh 输入框前添加元素
   * @en The label text displayed before (on the left side of) the input-tag field
   * @version 2.47.0
   */
  addBefore?: ReactNode;
  /**
   * @zh 输入框后添加元素
   * @en The label text displayed after (on the right side of) the input-tag field
   * @version 2.47.0
   */
  addAfter?: ReactNode;
  /**
   * @zh 自定义图标
   * @en Custom icons
   */
  icon?: { removeIcon?: ReactNode; clearIcon?: ReactNode };
  /**
   * @zh 触发自动分词的分隔符
   * @en Separator used to tokenize
   * @version 2.44.0
   */
  tokenSeparators?: string[];
  /**
   * @zh 校验函数，默认在 按下enter时候触发。
   * @en Function to check user's input, which is triggered when `Enter` is pressed
   * @defaultValue (inputValue, values) => inputValue && values.every((item) => item !== inputValue)
   * @version return type T and `Promise<T>` in 2.37.0
   */
  validate?: (inputValue: string, values: T[]) => boolean | Promise<boolean> | T | Promise<T>;
  /**
   * @zh 自定义标签渲染，`props` 为当前标签属性，`index` 为当前标签的顺序，`values` 为所有标签的值.
   * @en Custom tag rendering, `props` is the current tag attribute, `index` is the order of the current tag, `values` is the value of all tags
   * @version index、values added in 2.15.0
   */
  renderTag?: (
    props: {
      value: any;
      label: ReactNode;
      closable: boolean;
      onClose: (event) => void;
    },
    index: number,
    values: ObjectValueType[]
  ) => ReactNode;
  /**
   * @zh 移除某一个标签时改变时触发
   * @en Callback when a tag is removed
   */
  onRemove?: (value: T, index: number, event) => void;
  /**
   * @zh 控件值改变时触发
   * @en Callback when value changes
   * @version `reason` in 2.27.0
   */
  onChange?: (value: T[], reason: ValueChangeReason) => void;
  /**
   * @zh 失去焦点时候触发
   * @en Callback when input is blurred
   */
  onBlur?: (e) => void;
  /**
   * @zh 聚焦时触发
   * @en Callback when input is focused
   */
  onFocus?: (e) => void;
  /**
   * @zh 按 enter 键触发
   * @en Callback when `Enter` key is pressed
   */
  onPressEnter?: (e) => void;
  /**
   * @zh 输入框文本改变的回调。
   * @en Callback when the value of input changes
   */
  onInputChange?: (inputValue: string, event?) => void;
  /**
   * @zh 键盘事件回调
   * @en Callback when the keyboard is pressed
   */
  onKeyDown?: (e) => void;
  /**
   * @zh 输入框文本粘贴的回调。
   * @en Callback when you paste text in input box
   */
  onPaste?: (e) => void;
  /**
   * @zh 点击清除按钮的回调
   * @en Callback when clear button is clicked
   * @version 2.20.0
   */
  onClear?: () => void;
  /**
   * @zh 单击组件的回调。
   * @en Callback when the component is clicked
   */
  onClick?: (e) => void;
  tagClassName?: string;
  disableInput?: boolean;
}
