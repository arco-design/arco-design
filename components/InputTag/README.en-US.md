`````
Component / Data Entry

# InputTag

An input box which will display your input as tags.
`````

%%Content%%

## API

### InputTag

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|className|Additional css class|`string \| string[]`|`-`|-|
|style|Additional style|`CSSProperties`|`-`|-|
|size|Different sizes|`'mini' \| 'small' \| 'default' \| 'large'`|`-`|-|
|placeholder|Placeholder of input element|`string`|`-`|-|
|error|Error style|`boolean`|`-`|-|
|disabled|Whether the input is disabled|`boolean`|`-`|-|
|autoFocus|Whether the input is focused by default|`boolean`|`-`|-|
|readOnly|Whether the input is read only|`boolean`|`-`|-|
|allowClear|Whether to allow clear input value|`boolean`|`-`|-|
|animation|Whether to add animation for internal label changes|`boolean`|`true`|2.15.0|
|saveOnBlur|Whether to automatically store the text entering when blur InputTag|`boolean`|`-`|2.25.0|
|defaultValue|To set default value|`T[]`|`-`|-|
|value|To set value|`T[]`|`-`|-|
|inputValue|To set input value|`string`|`-`|-|
|labelInValue|If true, the incoming and callback values will be `{label: '', value: '')` format|`boolean`|`-`|-|
|dragToSort|Weather it is possible to sort tags by drag|`boolean`|`-`|2.27.0|
|suffix|The suffix for the InputTag|`React.ReactNode`|`-`|-|
|icon|Custom icons|`{ removeIcon?: ReactNode; clearIcon?: ReactNode }`|`-`|-|
|validate|Function to check user's input, which is triggered when `Enter` is pressed|`(inputValue: string, values: T[]) => boolean \| Promise<boolean>`|`(inputValue, values) => inputValue && values.every((item) => item !== inputValue)`|-|
|renderTag|Custom tag rendering, `props` is the current tag attribute, `index` is the order of the current tag, `values` is the value of all tags|`(props: {value: any;label: ReactNode;closable: boolean;onClose: (event) => void;},index: number,values: ObjectValueType[]) => React.ReactNode`|`-`|index、values added in 2.15.0|
|onRemove|Callback when a tag is removed|`(value: T, index: number, event) => void`|`-`|-|
|onChange|Callback when value changes|`(value: T[], reason: ValueChangeReason) => void`|`-`|`reason` in 2.27.0|
|onBlur|Callback when input is blurred|`(e) => void`|`-`|-|
|onFocus|Callback when input is focused|`(e) => void`|`-`|-|
|onPressEnter|Callback when `Enter` key is pressed|`(e) => void`|`-`|-|
|onInputChange|Callback when the value of input changes|`(inputValue: string, event?) => void`|`-`|-|
|onKeyDown|Callback when the keyboard is pressed|`(e) => void`|`-`|-|
|onPaste|Callback when you paste text in input box|`(e) => void`|`-`|-|
|onClear|Callback when clear button is clicked|`() => void`|`-`|2.20.0|
|onClick|Callback when the component is clicked|`(e) => void`|`-`|-|
