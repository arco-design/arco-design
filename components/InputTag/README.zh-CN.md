`````
组件 / 输入

# InputTag

标签输入。
`````

%%Content%%

## API

### InputTag

|参数名|描述|类型|默认值|版本|
|---|---|---|---|---|
|className|节点类名|`string \| string[]`|`-`|-|
|style|节点样式|`CSSProperties`|`-`|-|
|size|不同尺寸|`'mini' \| 'small' \| 'default' \| 'large'`|`-`|-|
|placeholder|预设文案|`string`|`-`|-|
|error|是否是错误状态|`boolean`|`-`|-|
|disabled|是否禁用|`boolean`|`-`|-|
|autoFocus|自动聚焦|`boolean`|`-`|-|
|readOnly|是否只读|`boolean`|`-`|-|
|allowClear|是否允许清除|`boolean`|`-`|-|
|animation|是否为内部标签变化添加动画。|`boolean`|`true`|2.15.0|
|saveOnBlur|是否在失焦时自动存储正在输入的文本|`boolean`|`-`|2.25.0|
|defaultValue|默认值|`T[]`|`-`|-|
|value|控件值|`T[]`|`-`|-|
|inputValue|控件的输入框内的值|`string`|`-`|-|
|labelInValue|设置传入和回调出的值均为 `{ label: '', value: ''}` 格式。|`boolean`|`-`|-|
|dragToSort|是否可以通过拖拽为 Tag 排序|`boolean`|`-`|2.27.0|
|suffix|后缀|`React.ReactNode`|`-`|-|
|icon|自定义图标|`{ removeIcon?: ReactNode; clearIcon?: ReactNode }`|`-`|-|
|validate|校验函数，默认在 按下enter时候触发。|`(inputValue: string, values: T[]) => boolean \| Promise<boolean>`|`(inputValue, values) => inputValue && values.every((item) => item !== inputValue)`|-|
|renderTag|自定义标签渲染，`props` 为当前标签属性，`index` 为当前标签的顺序，`values` 为所有标签的值.|`(props: {value: any;label: ReactNode;closable: boolean;onClose: (event) => void;},index: number,values: ObjectValueType[]) => React.ReactNode`|`-`|index、values added in 2.15.0|
|onRemove|移除某一个标签时改变时触发|`(value: T, index: number, event) => void`|`-`|-|
|onChange|控件值改变时触发|`(value: T[], reason: ValueChangeReason) => void`|`-`|`reason` in 2.27.0|
|onBlur|失去焦点时候触发|`(e) => void`|`-`|-|
|onFocus|聚焦时触发|`(e) => void`|`-`|-|
|onPressEnter|按 enter 键触发|`(e) => void`|`-`|-|
|onInputChange|输入框文本改变的回调。|`(inputValue: string, event?) => void`|`-`|-|
|onKeyDown|键盘事件回调|`(e) => void`|`-`|-|
|onPaste|输入框文本粘贴的回调。|`(e) => void`|`-`|-|
|onClear|点击清除按钮的回调|`() => void`|`-`|2.20.0|
|onClick|单击组件的回调。|`(e) => void`|`-`|-|
