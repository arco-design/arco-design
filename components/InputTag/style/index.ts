import '../../_class/OverflowEllipsis/style/index.less';
import '../../style/index.less';
import '../../Tag/style';
import './index.less';
