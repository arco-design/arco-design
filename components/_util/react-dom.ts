import { ReactElement } from 'react';
import ReactDOM from 'react-dom';
import { isObject } from './is';

type CreateRootFnType = (container: Element | DocumentFragment) => {
  render: (container: ReactElement) => void;
  unmount: () => void;
  _unmount: () => void;
};

const __SECRET_INTERNALS__ = '__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED';

const CopyReactDOM = ReactDOM as typeof ReactDOM & {
  createRoot: CreateRootFnType;
  // https://github.com/facebook/react/blob/4ff5f5719b348d9d8db14aaa49a48532defb4ab7/packages/react-dom/src/client/ReactDOM.js#L181
  [__SECRET_INTERNALS__]: {
    usingClientEntryPoint?: boolean;
  };
};

let copyRender: (
  app: ReactElement,
  container: Element | DocumentFragment
) => {
  render: (container: ReactElement) => void;
  _unmount: () => void;
};

const isReact18 = Number(CopyReactDOM.version?.split('.')[0]) > 17;

const updateUsingClientEntryPoint = (skipWarning?: boolean) => {
  // https://github.com/facebook/react/blob/17806594cc28284fe195f918e8d77de3516848ec/packages/react-dom/npm/client.js#L10
  // Avoid console warning
  if (isObject(CopyReactDOM[__SECRET_INTERNALS__])) {
    CopyReactDOM[__SECRET_INTERNALS__].usingClientEntryPoint = skipWarning;
  }
};

let createRoot: CreateRootFnType;
try {
  createRoot = CopyReactDOM.createRoot;
} catch (_) {
  //
}

if (isReact18 && createRoot) {
  copyRender = (app: ReactElement, container: Element | DocumentFragment) => {
    updateUsingClientEntryPoint(true);
    const root = createRoot(container);
    updateUsingClientEntryPoint(false);

    root.render(app);

    root._unmount = function () {
      setTimeout(() => {
        root?.unmount?.();
      });
    };
    return root;
  };
} else {
  copyRender = function (app: ReactElement, container: Element | DocumentFragment) {
    CopyReactDOM.render(app, container);

    return {
      render: (app: ReactElement) => {
        CopyReactDOM.render(app, container);
      },
      _unmount() {
        CopyReactDOM.unmountComponentAtNode(container);
      },
    };
  };
}

export const render = copyRender;
