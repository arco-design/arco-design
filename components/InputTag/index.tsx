import InputTag, { InputTagProps } from './input-tag';

export default InputTag;
export type { InputTagProps };
