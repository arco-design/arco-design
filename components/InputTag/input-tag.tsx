import React, {
  useContext,
  useState,
  useRef,
  useImperativeHandle,
  ElementRef,
  useEffect,
  PropsWithChildren,
  ReactNode,
} from 'react';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import { ConfigContext } from '../ConfigProvider';
import Tag from '../Tag';
import useMergeValue from '../_util/hooks/useMergeValue';
import cs from '../_util/classNames';
import InputComponent from '../Input/input-element';
import IconHover from '../_class/icon-hover';
import IconClose from '../../icon/react-icon/IconClose';
import { isObject, isArray } from '../_util/is';
import getHotkeyHandler from '../_util/getHotkeyHandler';
import { Backspace } from '../_util/keycode';
// import { pickTriggerPropsFromRest } from '../_util/constant';
import { ObjectValueType, InputTagProps, ValueChangeReason } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';
import Draggable from '../_class/Draggable';
import omit from '../_util/omit';
import fillNBSP from '../_util/fillNBSP';

const CSS_TRANSITION_DURATION = 300;
const REACT_KEY_FOR_INPUT = `__input_${Math.random().toFixed(10).slice(2)}`;

// 输入框粘贴会先触发 onPaste 后触发 onChange，但 onChange 的 value 中不包含换行符
// 如果刚刚因为粘贴触发过分词，则 onChange 不再进行分词尝试
const THRESHOLD_TOKEN_SEPARATOR_TRIGGER = 100;

const isEmptyNode = (node: ReactNode): boolean => {
  return node === null || node === undefined;
};

const keepFocus = (e) => {
  e.target.tagName !== 'INPUT' && e.preventDefault();
};

const formatValue = (value: (ObjectValueType | string)[]) => {
  if (!isArray(value)) {
    return [];
  }
  return value.map((item) => {
    return isObject(item)
      ? {
          ...item,
          label: 'label' in item ? item.label : item.value,
          value: item.value,
          closable: item.closable,
        }
      : {
          label: item,
          value: item,
        };
  });
};

type InputTagHandle = {
  focus: () => void;
  blur: () => void;
};

// Deal with the delay of recomputing input width
const useComputeAutoWidthDelay = (value: InputTagProps<any>['value']) => {
  const refDelay = useRef(0);
  const refPrevValueLength = useRef(value.length);

  useEffect(() => {
    refDelay.current =
      value.length === 0 && refPrevValueLength.current > 0 ? CSS_TRANSITION_DURATION : 0;
    refPrevValueLength.current = value.length;
  }, [value]);

  return refDelay;
};

const UsedTransitionGroup = ({
  prefixCls,
  children,
  animation,
}: PropsWithChildren<{ prefixCls: string; animation: boolean }>) => {
  return animation ? (
    <TransitionGroup component="div" className={`${prefixCls}-inner`}>
      {children}
    </TransitionGroup>
  ) : (
    <div className={`${prefixCls}-inner`}>{children}</div>
  );
};

const defaultProps: InputTagProps = {
  animation: true,
  validate: (inputValue, values) => inputValue && values.every((item) => item.value !== inputValue),
};

function InputTag(baseProps: InputTagProps<string | ObjectValueType>, ref) {
  const { getPrefixCls, size: ctxSize, componentConfig, rtl } = useContext(ConfigContext);
  const props = useMergeProps<InputTagProps>(baseProps, defaultProps, componentConfig?.InputTag);
  const {
    className,
    style,
    placeholder,
    error,
    disabled,
    readOnly,
    allowClear,
    autoFocus,
    labelInValue,
    disableInput,
    animation,
    saveOnBlur,
    dragToSort,
    icon,
    prefix,
    suffix,
    addBefore,
    addAfter,
    tokenSeparators,
    validate,
    renderTag,
    tagClassName,
    onInputChange,
    onKeyDown,
    onPaste,
    onChange,
    onFocus,
    onBlur,
    onPressEnter,
    onRemove,
    onClear,
    onClick,
    ...rest
  } = props;
  const prefixCls = getPrefixCls('input-tag');
  const size = 'size' in props ? props.size : ctxSize;

  const refInput = useRef<ElementRef<typeof InputComponent>>();
  const refTSLastSeparateTriggered = useRef<number>(null);

  const [focused, setFocused] = useState(false);
  const [value, setValue] = useMergeValue<ObjectValueType[]>([], {
    defaultValue: 'defaultValue' in props ? formatValue(props.defaultValue) : undefined,
    value: 'value' in props ? formatValue(props.value) : undefined,
  });
  const [inputValue, setInputValue] = useMergeValue('', {
    value: props.inputValue,
  });
  const refDelay = useComputeAutoWidthDelay(value);
  const draggable = !!(dragToSort && !readOnly && !disabled);

  useImperativeHandle<any, InputTagHandle>(
    ref,
    () => {
      return {
        blur: refInput.current?.blur,
        focus: refInput.current?.focus,
      };
    },
    []
  );

  const valueChangeHandler = (value: ObjectValueType[], reason: ValueChangeReason) => {
    if (disabled || readOnly) {
      return;
    }
    if (!('value' in props)) {
      setValue(value);
    }

    onChange && onChange(labelInValue ? value : value.map((x) => x.value), reason);
  };

  const tagCloseHandler = (itemValue: ObjectValueType, index: number, event) => {
    onRemove && onRemove(itemValue, index, event);
    valueChangeHandler([...value.slice(0, index), ...value.slice(index + 1)], 'remove');
  };

  const hotkeyHandler = getHotkeyHandler(
    new Map([
      [
        Backspace.code,
        (event) => {
          if (!event.target.value && value.length) {
            for (let index = value.length - 1; index >= 0; index--) {
              const itemValue = value[index];
              if (itemValue.closable !== false) {
                tagCloseHandler(itemValue, index, event);
                return;
              }
            }
          }
        },
      ],
    ])
  );

  const tryAddInputValueToTag = async () => {
    try {
      const validateResult =
        typeof validate === 'function' ? await validate(inputValue, value) : true;
      if (validateResult) {
        valueChangeHandler(
          value.concat({
            value: validateResult === true ? inputValue : validateResult,
            label: inputValue,
          }),
          'add'
        );
        setInputValue('');
      }
    } catch (error) {
      console.error(error);
    }
  };

  const mergedRenderTag = (item: ObjectValueType, index: number) => {
    const { value: itemValue, label } = item;
    const closable = !readOnly && !disabled && item.closable !== false;
    const onClose = (event) => {
      tagCloseHandler(item, index, event);
    };

    if (renderTag) {
      return renderTag(
        {
          value: itemValue,
          label,
          closable,
          onClose,
        },
        index,
        value
      );
    }

    return (
      <Tag
        visible
        className={cs(`${prefixCls}-tag`, {
          [tagClassName]: tagClassName,
        })}
        closable={closable}
        closeIcon={icon && icon.removeIcon}
        __closeIconProps={{
          onMouseDown: keepFocus,
        }}
        title={typeof label === 'string' ? label : undefined}
        onClose={onClose}
      >
        {fillNBSP(label)}
      </Tag>
    );
  };

  const handleTokenSeparators = async (str: string, isPaste = false) => {
    // won't handle token separators in a short time
    if (
      isPaste &&
      Date.now() - refTSLastSeparateTriggered.current < THRESHOLD_TOKEN_SEPARATOR_TRIGGER
    ) {
      return;
    }

    // clear the timestamp, and then we can judge whether tokenSeparators has been triggered
    // according to timestamp value
    refTSLastSeparateTriggered.current = null;

    if (isArray(tokenSeparators) && tokenSeparators.length) {
      const splitTextList = str.split(new RegExp(`[${tokenSeparators.join('')}]`));

      if (splitTextList.length > 1) {
        // record the timestamp of tokenSeparators triggered
        refTSLastSeparateTriggered.current = Date.now();

        const validatedValueList: ObjectValueType[] = [];

        await Promise.all(
          splitTextList.map(async (text) => {
            // filter empty string and validate it
            const validateResult = text
              ? typeof validate === 'function'
                ? await validate(text, value)
                : true
              : false;
            if (validateResult) {
              validatedValueList.push({
                value: validateResult === true ? text : validateResult,
                label: text,
              });
            }
          })
        );

        if (validatedValueList.length) {
          valueChangeHandler(value.concat(validatedValueList), 'add');
        }
      }
    }
  };

  const clearIcon =
    allowClear && !disabled && !readOnly && value.length ? (
      <IconHover
        size={size}
        key="clearIcon"
        className={`${prefixCls}-clear-icon`}
        onClick={(e) => {
          e.stopPropagation();
          valueChangeHandler([], 'clear');
          if (!focused) {
            refInput.current?.focus();
          }
          onClear && onClear();
        }}
      >
        {(icon && icon.clearIcon) || <IconClose />}
      </IconHover>
    ) : null;

  const disableInputComponent = disabled || disableInput;

  // CSSTransition needs to be a direct child of TransitionGroup, otherwise the animation will NOT work
  // https://github.com/arco-design/arco-design/issues/622
  const childrenWithAnimation = value
    .map((x, i) => {
      // Check whether two tags have same value. If so, set different key for them to avoid only rendering one tag.
      const isRepeat = value.findIndex((item) => item.value === x.value) !== i;
      const eleTag = mergedRenderTag(x, i);
      return React.isValidElement(eleTag) ? (
        <CSSTransition
          key={typeof x.value === 'object' ? i : isRepeat ? `${x.value}-${i}` : x.value}
          timeout={CSS_TRANSITION_DURATION}
          classNames="zoomIn"
        >
          {eleTag}
        </CSSTransition>
      ) : (
        eleTag
      );
    })
    .concat(
      <CSSTransition
        key={REACT_KEY_FOR_INPUT}
        timeout={CSS_TRANSITION_DURATION}
        classNames="zoomIn"
      >
        <InputComponent
          autoComplete="off"
          size={size}
          disabled={disableInputComponent}
          readOnly={readOnly}
          ref={refInput}
          autoFocus={autoFocus}
          placeholder={!value.length ? placeholder : ''}
          prefixCls={`${prefixCls}-input`}
          autoFitWidth={{
            delay: () => refDelay.current,
          }}
          onPressEnter={async (e) => {
            inputValue && e.preventDefault();
            onPressEnter?.(e);
            await tryAddInputValueToTag();
          }}
          onFocus={(e) => {
            if (!disableInputComponent && !readOnly) {
              setFocused(true);
              onFocus?.(e);
            }
          }}
          onBlur={async (e) => {
            setFocused(false);
            onBlur?.(e);
            if (saveOnBlur) {
              await tryAddInputValueToTag();
            }
            setInputValue('');
          }}
          value={inputValue}
          onChange={(value, event) => {
            // Only fire callback on user input to ensure parent component can get real input value on controlled mode.
            onInputChange?.(value, event);

            const inputType = event.nativeEvent.inputType;
            // do NOT use await, need to update input value right away
            handleTokenSeparators(value, inputType === 'insertFromPaste');

            if (refTSLastSeparateTriggered.current) {
              setInputValue('');
            } else {
              setInputValue(value);
            }
          }}
          onKeyDown={(event) => {
            hotkeyHandler(event as any);
            onKeyDown?.(event);
          }}
          onPaste={(event) => {
            onPaste?.(event);
            handleTokenSeparators(event.clipboardData.getData('text'));
          }}
        />
      </CSSTransition>
    );

  const hasPrefix = !isEmptyNode(prefix);
  const hasSuffix = !isEmptyNode(suffix) || !isEmptyNode(clearIcon);
  const needAddBefore = !isEmptyNode(addBefore);
  const needAddAfter = !isEmptyNode(addAfter);
  const needWrapper = needAddBefore || needAddAfter;

  const status = props.status || (error ? 'error' : undefined);
  const innerClassNames = cs(prefixCls, {
    [`${prefixCls}-size-${size}`]: size,
    [`${prefixCls}-disabled`]: disabled,
    [`${prefixCls}-${status}`]: status,
    [`${prefixCls}-focus`]: focused,
    [`${prefixCls}-readonly`]: readOnly,
    [`${prefixCls}-has-suffix`]: hasSuffix,
    [`${prefixCls}-has-placeholder`]: !value.length,
    [`${prefixCls}-rtl`]: rtl,
  });
  const propsAppliedToRoot = { style, className };

  const eleInputTagCore = (
    <div
      {...omit(rest, ['status', 'size', 'defaultValue', 'value', 'inputValue'])}
      {...(needWrapper ? {} : propsAppliedToRoot)}
      className={needWrapper ? innerClassNames : cs(innerClassNames, propsAppliedToRoot.className)}
      onMouseDown={(event) => {
        focused && keepFocus(event);
      }}
      onClick={(e) => {
        !focused && refInput.current?.focus();
        if (onClick) {
          onClick(e);
        }
      }}
    >
      <div className={`${prefixCls}-view`}>
        {hasPrefix && (
          <div className={`${prefixCls}-prefix`} onMouseDown={keepFocus}>
            {prefix}
          </div>
        )}

        {draggable ? (
          <UsedTransitionGroup
            key="transitionGroupWithDrag"
            prefixCls={prefixCls}
            animation={animation}
          >
            <Draggable
              itemWrapperStyle={{ display: 'inline-block' }}
              direction="horizontal"
              onIndexChange={(index, prevIndex) => {
                const moveItem = function (arr, fromIndex, toIndex) {
                  arr = arr.slice();
                  const isMoveLeft = fromIndex > toIndex;
                  const [item] = arr.splice(fromIndex, 1);
                  arr.splice(isMoveLeft ? toIndex : toIndex - 1, 0, item);
                  return arr;
                };
                valueChangeHandler(moveItem(value, prevIndex, index), 'sort');
              }}
            >
              {childrenWithAnimation}
            </Draggable>
          </UsedTransitionGroup>
        ) : (
          <UsedTransitionGroup prefixCls={prefixCls} animation={animation}>
            {childrenWithAnimation}
          </UsedTransitionGroup>
        )}

        {hasSuffix && (
          <div className={`${prefixCls}-suffix`} onMouseDown={keepFocus}>
            {clearIcon}
            {suffix}
          </div>
        )}
      </div>
    </div>
  );

  if (!needWrapper) {
    return eleInputTagCore;
  }

  return (
    <div
      {...propsAppliedToRoot}
      className={cs(
        `${prefixCls}-wrapper`,
        {
          [`${prefixCls}-wrapper-rtl`]: rtl,
        },
        propsAppliedToRoot.className
      )}
    >
      {needAddBefore && <div className={`${prefixCls}-addbefore`}>{addBefore}</div>}
      {eleInputTagCore}
      {needAddAfter && <div className={`${prefixCls}-addafter`}>{addAfter}</div>}
    </div>
  );
}

const InputTagRef = React.forwardRef<InputTagHandle, InputTagProps<any>>(InputTag);

InputTagRef.displayName = 'InputTag';

export default InputTagRef;

export { InputTagProps };
