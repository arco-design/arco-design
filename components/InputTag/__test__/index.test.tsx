import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import InputTag from '..';
import { sleep } from '../../../tests/util';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import { Backspace, Enter } from '../../_util/keycode';

mountTest(InputTag);
componentConfigTest(InputTag, 'InputTag');

const getInputAttr = (wrapper, name) => {
  return wrapper.find('input').getDOMNode().getAttribute(name);
};

describe('InputTag', () => {
  it('render correctly', () => {
    const wrapper = mount(<InputTag placeholder="请输入" />);

    expect(getInputAttr(wrapper, 'placeholder')).toEqual('请输入');

    wrapper.setProps({
      readOnly: true,
    });

    expect(getInputAttr(wrapper, 'readonly')).toBe('');
    wrapper.setProps({
      disabled: true,
    });

    expect(getInputAttr(wrapper, 'disabled')).toEqual('');

    wrapper.setProps({
      error: true,
    });

    expect(wrapper.find('.arco-input-tag').hasClass('arco-input-tag-error')).toBe(true);
  });

  it('input correctly', async () => {
    const mockOnChange = jest.fn();
    const mockOnRemove = jest.fn();
    const wrapper = mount(
      <InputTag
        onChange={mockOnChange}
        onRemove={mockOnRemove}
        placeholder="请输入"
        defaultValue={['aaa']}
      />
    );

    expect(wrapper.find('Tag')).toHaveLength(1);

    // 聚焦
    act(() => {
      wrapper.find('input').simulate('focus');
    });
    wrapper.update();
    expect(wrapper.find('.arco-input-tag').hasClass('arco-input-tag-focus')).toBe(true);

    // 输入后回车
    act(() => {
      wrapper
        .find('input')
        .simulate('change', { currentTarget: { value: 'bbb' }, target: { value: 'bbb' } });
    });
    // 回车依赖了输入的值，所以要放在下一个 act 中
    act(() => {
      wrapper.find('input').simulate('keydown', { keyCode: Enter.code });
    });
    await sleep(10);
    expect(mockOnChange.mock.calls.length).toBe(1);

    // 删除
    act(() => {
      wrapper.find('input').simulate('keydown', { keyCode: Backspace.code });
    });
    expect(mockOnRemove.mock.calls.length).toBe(1);
  });

  it('saveOnBlur works', async () => {
    const inputValue = 'bbb';
    const mockOnChange = jest.fn();
    const wrapper = mount(<InputTag saveOnBlur onChange={mockOnChange} />);
    act(() => {
      wrapper.find('input').simulate('change', {
        currentTarget: { value: inputValue },
        target: { value: inputValue },
      });
    });
    act(() => {
      wrapper.find('input').simulate('blur');
    });
    await sleep(10);
    expect(JSON.stringify(mockOnChange.mock.calls[0][0])).toBe(JSON.stringify([inputValue]));
  });

  it('onClick called correctly', async () => {
    const mockOnClick = jest.fn();
    const wrapper = mount(<InputTag onClick={mockOnClick} />);

    // click
    act(() => {
      wrapper.find('input').simulate('click');
    });
    expect(mockOnClick).toBeCalled();
  });
});
