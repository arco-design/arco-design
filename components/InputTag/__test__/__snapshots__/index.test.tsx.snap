// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.InputTag default 1`] = `
<div
  class="arco-input-tag arco-input-tag-size-default arco-input-tag-has-placeholder"
>
  <div
    class="arco-input-tag-view"
  >
    <div
      class="arco-input-tag-inner"
    >
      <input
        autocomplete="off"
        class="arco-input-tag-input arco-input-tag-input-size-default"
        value=""
      />
      <span
        class="arco-input-tag-input-mirror"
      />
    </div>
  </div>
</div>
`;

exports[`ConfigProvider componentConfig.InputTag set className globally 1`] = `
<div
  class="arco-input-tag arco-input-tag-size-default arco-input-tag-has-placeholder global-default-class"
>
  <div
    class="arco-input-tag-view"
  >
    <div
      class="arco-input-tag-inner"
    >
      <input
        autocomplete="off"
        class="arco-input-tag-input arco-input-tag-input-size-default"
        value=""
      />
      <span
        class="arco-input-tag-input-mirror"
      />
    </div>
  </div>
</div>
`;

exports[`ConfigProvider componentConfig.InputTag set className globally and component set className itself. 1`] = `
<div
  class="arco-input-tag arco-input-tag-size-default arco-input-tag-has-placeholder component-class"
>
  <div
    class="arco-input-tag-view"
  >
    <div
      class="arco-input-tag-inner"
    >
      <input
        autocomplete="off"
        class="arco-input-tag-input arco-input-tag-input-size-default"
        value=""
      />
      <span
        class="arco-input-tag-input-mirror"
      />
    </div>
  </div>
</div>
`;

exports[`ConfigProvider componentConfig.InputTag set data-* property 1`] = `
<div
  class="arco-input-tag arco-input-tag-size-default arco-input-tag-has-placeholder"
  data-test-name="InputTag"
>
  <div
    class="arco-input-tag-view"
  >
    <div
      class="arco-input-tag-inner"
    >
      <input
        autocomplete="off"
        class="arco-input-tag-input arco-input-tag-input-size-default"
        value=""
      />
      <span
        class="arco-input-tag-input-mirror"
      />
    </div>
  </div>
</div>
`;
